src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java

```diff
@@ -136,6 +155,83 @@ private MissingEnvironment(Label environment, RemovedEnvironmentCulprit culprit)
         continue; // This target doesn't participate in constraints.
       }
 
+      Object target_compatible_with_object =
+          target.getAssociatedRule().getAttributeContainer().getAttr("target_compatible_with");
+
+      // Get the target platform key.
+      ConfiguredTargetKey targetPlatformKey =
```

Comment:
While I don't think there's much cost, I'd consider not even getting the platform for the majority of targets that don't have `target_compatible_with` set to anything.

Reply:
Good point. Will do.